// amexio-chart-d3/esm2015/lib/legend/legend.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from "@angular/core";
export class AmexioD3Legend {
    constructor() {
        this.horizontal = false;
        this.onClick = new EventEmitter();
    }
    /**
     * @param {?} node
     * @return {?}
     */
    onLegendClick(node) {
        this.onClick.emit(node);
    }
}
AmexioD3Legend.decorators = [
    { type: Component, args: [{
                selector: 'amexio-d3-legend',
                template: `<div style =" width: 100% ; padding-left: 20px; padding-top: 8px; ">
<table *ngIf="!horizontal" style="height: 100%;">
    <tr *ngFor = "let node of data" style="cursor: pointer;" (click)="onLegendClick(node)">
        <td style="height: 10px; width: 23px; padding-left: 2px;"
            [ngStyle]="{'background-color': node.color}"></td>
        <td>{{node.label}}</td>
    </tr>
</table>
</div>

<table *ngIf="horizontal" style="height: 100%;">
    <tr>
        <td *ngFor = "let node of data" 
                    (click)="onLegendClick(node)" 
                    style="cursor: pointer;">
            
            <div style=" display: inline-block; display: inline-block; padding: 2px;">
                    <div style=" width: 10px; display: inline-block; padding: 2px; " 
                    [ngStyle]="{'background-color': node.color}"></div>
                {{node.label}}
            </div>
        </td>
        
    </tr>
</table>`
            },] },
];
AmexioD3Legend.ctorParameters = () => [];
AmexioD3Legend.propDecorators = {
    data: [{ type: Input, args: ['data',] }],
    horizontal: [{ type: Input, args: ['horizontal',] }],
    onClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    AmexioD3Legend.prototype.data;
    /** @type {?} */
    AmexioD3Legend.prototype.horizontal;
    /** @type {?} */
    AmexioD3Legend.prototype.onClick;
}
