// amexio-chart-d3/esm2015/lib/services/comman.data.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class CommanDataService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @param {?} serviceUrl
     * @param {?} methodType
     * @return {?}
     */
    fetchUrlData(serviceUrl, methodType) {
        /** @type {?} */
        const requestJson = {};
        /** @type {?} */
        const headers = new HttpHeaders().append('Content-Type', 'application/json;charset=UTF-8');
        if (methodType === 'post') {
            return this.http.post(serviceUrl, requestJson, { headers });
        }
        else if (methodType === 'get') {
            return this.http.get(serviceUrl, { headers });
        }
    }
    /**
     * @param {?} serviceUrl
     * @param {?} methodType
     * @param {?} requestJson
     * @return {?}
     */
    postfetchData(serviceUrl, methodType, requestJson) {
        /** @type {?} */
        const headers = new HttpHeaders().append('Content-Type', 'application/json;charset=UTF-8');
        if (methodType === 'post') {
            return this.http.post(serviceUrl, requestJson, { headers });
        }
        else if (methodType === 'get') {
            return this.http.get(serviceUrl, { headers });
        }
    }
}
CommanDataService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
CommanDataService.ctorParameters = () => [
    { type: HttpClient }
];
/** @nocollapse */ CommanDataService.ngInjectableDef = i0.defineInjectable({ factory: function CommanDataService_Factory() { return new CommanDataService(i0.inject(i1.HttpClient)); }, token: CommanDataService, providedIn: "root" });
if (false) {
    /** @type {?} */
    CommanDataService.prototype.serviceUrl;
    /** @type {?} */
    CommanDataService.prototype.http;
}
