// amexio-chart-d3/esm2015/lib/barstack/barstack.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, Output, ViewChild, ElementRef, EventEmitter } from '@angular/core';
import * as d3 from 'd3';
import { AmexioD3BaseChartComponent } from '../base/base.component';
import { CommanDataService } from '../services/comman.data.service';
import { DeviceQueryService } from '../services/device.query.service';
export class BarstackComponent extends AmexioD3BaseChartComponent {
    /**
     * @param {?} myservice
     * @param {?} device
     */
    constructor(myservice, device) {
        super('barstack');
        this.myservice = myservice;
        this.device = device;
        this.maxYValue = 0;
        this.barwidth = 0;
        this.labelflag = false;
        this.title = "";
        this.legend = true;
        this.labelcolor = "black";
        this.color = [];
        this.svgwidth = 300;
        this.level = 0;
        this.drillabledatakey = [];
        this.hScale = true;
        this.drillableFlag = true;
        this.resizeflag = false;
        this.svgheight = 300;
        this.onLegendClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.level <= 1) {
            /** @type {?} */
            let res;
            if (this.httpmethod && this.httpurl) {
                this.myservice.fetchUrlData(this.httpurl, this.httpmethod).subscribe((response) => {
                    //this.data = response;
                    this.httpresponse = response;
                    this.data = this.getResponseData(response);
                }, (error) => {
                }, () => {
                    setTimeout(() => {
                        this.transformData(this.data);
                        this.plotChart();
                    }, 0);
                });
            }
            else if (this.data1) {
                setTimeout(() => {
                    this.transformData(this.data1);
                    this.plotChart();
                }, 0);
            }
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    fetchData(data) {
        /** @type {?} */
        let requestJson;
        /** @type {?} */
        let key = this.drillabledatakey;
        /** @type {?} */
        let resp;
        if (this.drillabledatakey.length) {
            /** @type {?} */
            let drillabledata = this.getMultipleDrillbleKeyData(data, key);
            requestJson = drillabledata;
        }
        else {
            requestJson = data;
        }
        if (this.httpmethod && this.httpurl) {
            this.myservice.postfetchData(this.httpurl, this.httpmethod, requestJson).subscribe((response) => {
                resp = response;
                this.httpresponse = response;
            }, (error) => {
            }, () => {
                setTimeout(() => {
                    //this.data = this.getResponseData(resp);
                    this.drawChart();
                }, 0);
            });
        }
    }
    /**
     * @return {?}
     */
    drawChart() {
        setTimeout(() => {
            this.data = this.getResponseData(this.httpresponse);
            this.transformData(this.data);
            this.plotChart();
        }, 0);
    }
    /**
     * @param {?} data1
     * @return {?}
     */
    transformData(data1) {
        this.keyArray = [];
        this.legendArray = [];
        data1.forEach((element, i) => {
            if (i == 0) {
                element.forEach((innerelement, index) => {
                    if (index > 0) {
                        this.legendArray[innerelement] = { 'data': [] };
                        this.keyArray.push(innerelement);
                    }
                    else if (index == 0) {
                        this.xaxis = innerelement;
                    }
                });
            }
        });
        /** @type {?} */
        let tempinnerarray;
        tempinnerarray = [];
        data1.forEach((element, index) => {
            if (index > 0) {
                /** @type {?} */
                let obj = {};
                element.forEach((innerelement, innerindex) => {
                    if (innerindex >= 0) {
                        /** @type {?} */
                        const key = this.keyArray[innerindex - 1];
                        obj[key] = element[innerindex];
                        /** @type {?} */
                        const legenddata = this.legendArray[key];
                        if (legenddata) {
                            legenddata.data.push({ 'value': element[innerindex], 'label': element[0] });
                        }
                    }
                });
                tempinnerarray.push(obj);
            }
        });
        this.data = [];
        tempinnerarray.forEach(element => {
            this.data.push(element);
        });
        /** @type {?} */
        let maxY = 0;
        /** @type {?} */
        let yaxismaxArray = [];
        //find max for yaxis
        this.data.forEach((element) => {
            for (let [key, value] of Object.entries(element)) {
                this.keyArray.forEach(key1 => {
                    if (key == key1) {
                        maxY = maxY + value;
                    }
                }); //keyarray loop ends here
            } //for ends here
            yaxismaxArray.push(maxY);
            maxY = 0;
        }); // foreach ends
        // foreach ends
        /** @type {?} */
        let tempLarge = 0;
        /** @type {?} */
        let i;
        for (i = 0; i < yaxismaxArray.length; i++) {
            if (yaxismaxArray[i] > tempLarge) {
                this.maxYValue = yaxismaxArray[i];
            } //if ends
        } // for ends
        this.legends = [];
        this.keyArray.forEach((element, index) => {
            /** @type {?} */
            const legenddata = this.legendArray[element];
            if (this.color.length > 0) {
                /** @type {?} */
                let object = { 'label': element, 'color': this.color[index], 'data': legenddata.data };
                this.legends.push(object);
            }
            else {
                /** @type {?} */
                let object = { 'label': element, 'color': this.predefinedcolors[index], 'data': legenddata.data };
                this.legends.push(object);
            }
        });
    }
    /**
     * @return {?}
     */
    plotChart() {
        /** @type {?} */
        const tooltip = this.toolTip(d3);
        /** @type {?} */
        let margin = { top: 20, right: 30, bottom: 90, left: 60 };
        /** @type {?} */
        let colors = this.predefinedcolors;
        if (this.device.IsDesktop()) {
            if (this.chartId) {
                this.svgwidth = this.chartId.nativeElement.offsetWidth;
            }
            else {
                this.svgwidth = this.svgwidth;
            }
        }
        //this.svgwidth = this.chartId.nativeElement.offsetWidth;
        /** @type {?} */
        let data;
        data = this.data;
        /** @type {?} */
        let keysetarray = [];
        if (this.httpmethod && this.httpurl) {
            for (let [key, value] of Object.entries(this.data[0])) {
                keysetarray.push(key);
            }
            this.keyArray = keysetarray;
            this.keyArray.splice(0, 1);
        }
        /** @type {?} */
        let series = d3.stack().keys(this.keyArray)
            .offset(d3.stackOffsetDiverging)(this.data);
        series;
        this.svg = d3.select("#" + this.componentId);
        /** @type {?} */
        let width = this.svgwidth - margin.left - margin.right;
        /** @type {?} */
        let height;
        // height=this.chartId.nativeElement.offsetHeight-20;
        if (this.device.IsDesktop()) {
            this.offsetheight = this.chartId.nativeElement.offsetHeight;
            height = this.offsetheight;
        }
        else {
            height = this.chartId.nativeElement.offsetHeight - 10;
        }
        /** @type {?} */
        let x = d3.scaleBand()
            .domain(data.map((d) => {
            return d[Object.keys(d)[0]];
        }))
            .rangeRound([margin.left, width - margin.right])
            .padding(0.35);
        /** @type {?} */
        let y = d3.scaleLinear()
            .domain([d3.min(this.stackMin(series)),
            this.maxYValue
            // d3.max(this.stackMax(series))
        ])
            .rangeRound([height - margin.bottom, margin.top]);
        // if (this.barwidth > 0) {
        //   this.barwidth = this.barwidth;
        // }
        // else {
        //   this.barwidth = x.bandwidth();
        // }
        if (this.device.IsDesktop() == true) {
            this.svg.append("g")
                .attr("transform", "translate(0," + y(0) + ")")
                .call(d3.axisBottom(x));
        }
        else {
            this.svg.append("g")
                .attr("transform", "translate(0," + y(0) + ")")
                .call(d3.axisBottom(x)).
                selectAll("text")
                .attr("y", 0)
                .attr("x", 9)
                .attr("dy", ".35em")
                .attr("transform", "rotate(60)")
                .style("text-anchor", "start");
        }
        this.svg.append("g")
            .attr("transform", "translate(" + margin.left + ",0)")
            .call(d3.axisLeft(y));
        this.plotLine(this.svg, x, y, height, width, margin.left);
        /** @type {?} */
        let svgRect = this.svg.append("g")
            .selectAll("g")
            .data(series)
            .enter().append("g")
            .attr("fill", (d, index) => {
            if (this.color.length > 0) {
                if (this.color[index]) {
                    return this.color[index];
                }
                else {
                    return colors[index];
                }
            }
            else {
                return colors[index];
            }
        })
            .selectAll("rect")
            .data((d) => {
            return d;
        });
        svgRect.enter().append("rect")
            .attr("width", x.bandwidth()).attr('id', (d, i) => {
            return d.data[i];
        })
            .attr("x", (d) => {
            return x(+d.data[Object.keys(d.data)[0]]);
        })
            .attr("y", (d, index) => {
            return y(d[1]);
        })
            .attr("cursor", "pointer")
            .attr("height", (d, index) => {
            return y(d[0]) - y(d[1]);
        })
            .on("mouseover", (d) => {
            return tooltip.style("visibility", "visible");
        })
            .on("mousemove", (d) => {
            return tooltip.html(this.setKey(d))
                .style("top", (d3.event.pageY - 10) + "px")
                .style("left", (d3.event.pageX + 10) + "px");
        })
            .on("mouseout", (d) => {
            return tooltip.style("visibility", "hidden");
        })
            .on("click", (d) => {
            this.setBarClickText(d);
            this.fordrillableClick(this, d, event);
            return tooltip.style("visibility", "hidden");
            // this.chartClick(d);
        });
        // -------------------------
        if (this.labelflag) {
            svgRect.enter()
                .append("text")
                .style("font-weight", "bold")
                .style("font-size", "1vw")
                .attr("text-anchor", "middle")
                .attr("fill", (d) => {
                if (this.labelcolor.length > 0) {
                    return this.labelcolor;
                }
                else {
                    return "black";
                }
            })
                .attr("x", (d) => {
                return x(+d.data[Object.keys(d.data)[0]]) + x.bandwidth() / 2;
                // +  margin.left;
            })
                .attr("y", (d, index) => {
                return y(d[1]) + 20;
            })
                .text(function (d) {
                return d[Object.keys(d)[1]] - d[Object.keys(d)[0]];
            });
        }
    }
    /**
     * @param {?} serie
     * @return {?}
     */
    stackMin(serie) {
        return d3.min(serie, function (d) { return d[0]; });
    }
    // stackMax(serie) {
    //   return d3.max(serie, function (d) { return d[1]; });
    // }
    /**
     * @return {?}
     */
    resize() {
        this.svgwidth = 0;
        this.svg.selectAll("*").remove();
        this.resizeflag = true;
        this.svgwidth = this.divid.nativeElement.offsetWidth;
        this.plotChart();
    }
    /**
     * @param {?} svg
     * @param {?} x
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @param {?} m
     * @return {?}
     */
    plotLine(svg, x, y, height, width, m) {
        if (this.hScale) {
            svg.append('g')
                .attr("transform", "translate(" + m + ",0)")
                .attr("color", "lightgrey")
                .call(d3.axisLeft(y)
                .tickSize(-width).tickFormat(''));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    legendClick(event) {
        /** @type {?} */
        let obj = {};
        obj["label"] = event.label;
        /** @type {?} */
        let data = [];
        event.data.forEach(element => {
            /** @type {?} */
            let object = {};
            object[element.label] = element.value;
            data.push(object);
        });
        obj["data"] = data;
        this.onLegendClick.emit(obj);
    }
    /**
     * @param {?} d
     * @return {?}
     */
    setKey(d) {
        /** @type {?} */
        let diff = d[0] - d[1];
        if (diff < 0) {
            diff = (diff * (-1));
        }
        for (let [key, value] of Object.entries(d.data)) {
            if (value == diff) {
                //  let object = { 'label': key, 'value': value , 'legend': d.data[Object.keys(d.data)[0]]};
                /** @type {?} */
                let object = {};
                object[key] = value;
                object[this.xaxis] = d.data[Object.keys(d.data)[0]];
                return (this.toolTipForBar(object));
            }
        }
    }
    /**
     * @param {?} d
     * @return {?}
     */
    setBarClickText(d) {
        /** @type {?} */
        let diff = d[0] - d[1];
        if (diff < 0) {
            diff = (diff * (-1));
        }
        // let object = {};
        /** @type {?} */
        let object = {};
        for (let [key, value] of Object.entries(d.data)) {
            if (value == diff) {
                object[key] = value;
                object[this.xaxis] = d.data[Object.keys(d.data)[0]];
            }
        }
        this.chartClick(object);
    }
    /**
     * @param {?} httpResponse
     * @return {?}
     */
    getResponseData(httpResponse) {
        /** @type {?} */
        let responsedata = httpResponse;
        if (this.datareader != null) {
            /** @type {?} */
            const dr = this.datareader.split('.');
            for (const ir of dr) {
                responsedata = responsedata[ir];
            }
        }
        else {
            responsedata = httpResponse;
        }
        return responsedata;
    }
}
BarstackComponent.decorators = [
    { type: Component, args: [{
                selector: 'amexio-d3-chart-barstack',
                template: `<div   *ngIf="drillableFlag" #divid (window:resize)="resize()">
<table style="width: 100%;" #chartId  (window:resize)="resize()">
  <tr><td style="font-size: 18px; font-weight: bold; "> {{title}} </td></tr>
  <tr>
      <td>
          <svg [attr.height]="svgheight" [attr.width]="svgwidth" [attr.id]="componentId">
          </svg>            
      </td>
  </tr>
  <tr>
      <td style="display: block; padding: 5px;">
              <amexio-d3-legend *ngIf="legend" [horizontal]="true" (onClick)="legendClick($event)" 
                                  [data]="legends" style="height:100%;">
              </amexio-d3-legend>
      </td>
  
  </tr>
</table>
</div>`,
                styles: [``]
            },] },
];
BarstackComponent.ctorParameters = () => [
    { type: CommanDataService },
    { type: DeviceQueryService }
];
BarstackComponent.propDecorators = {
    data1: [{ type: Input, args: ['data',] }],
    barwidth: [{ type: Input }],
    labelflag: [{ type: Input, args: ['label',] }],
    title: [{ type: Input }],
    legend: [{ type: Input }],
    labelcolor: [{ type: Input, args: ['label-color',] }],
    color: [{ type: Input }],
    svgwidth: [{ type: Input, args: ['width',] }],
    datareader: [{ type: Input, args: ['data-reader',] }],
    level: [{ type: Input, args: ['level',] }],
    target: [{ type: Input, args: ['target',] }],
    drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
    hScale: [{ type: Input, args: ['horizontal-scale',] }],
    svgheight: [{ type: Input, args: ['height',] }],
    chartId: [{ type: ViewChild, args: ['chartId',] }],
    divid: [{ type: ViewChild, args: ['divid',] }],
    drillid: [{ type: ViewChild, args: ['drillid',] }],
    onLegendClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    BarstackComponent.prototype.legendArray;
    /** @type {?} */
    BarstackComponent.prototype.maxYValue;
    /** @type {?} */
    BarstackComponent.prototype.keyArray;
    /** @type {?} */
    BarstackComponent.prototype.predefinedcolors;
    /** @type {?} */
    BarstackComponent.prototype.legends;
    /** @type {?} */
    BarstackComponent.prototype.charttype;
    /** @type {?} */
    BarstackComponent.prototype.data;
    /** @type {?} */
    BarstackComponent.prototype.datareaderdata;
    /** @type {?} */
    BarstackComponent.prototype.xaxis;
    /** @type {?} */
    BarstackComponent.prototype.data1;
    /** @type {?} */
    BarstackComponent.prototype.barwidth;
    /** @type {?} */
    BarstackComponent.prototype.labelflag;
    /** @type {?} */
    BarstackComponent.prototype.title;
    /** @type {?} */
    BarstackComponent.prototype.legend;
    /** @type {?} */
    BarstackComponent.prototype.labelcolor;
    /** @type {?} */
    BarstackComponent.prototype.color;
    /** @type {?} */
    BarstackComponent.prototype.svgwidth;
    /** @type {?} */
    BarstackComponent.prototype.datareader;
    /** @type {?} */
    BarstackComponent.prototype.level;
    /** @type {?} */
    BarstackComponent.prototype.target;
    /** @type {?} */
    BarstackComponent.prototype.drillabledatakey;
    /** @type {?} */
    BarstackComponent.prototype.hScale;
    /** @type {?} */
    BarstackComponent.prototype.drillableFlag;
    /** @type {?} */
    BarstackComponent.prototype.resizeflag;
    /** @type {?} */
    BarstackComponent.prototype.svgheight;
    /** @type {?} */
    BarstackComponent.prototype.chartId;
    /** @type {?} */
    BarstackComponent.prototype.divid;
    /** @type {?} */
    BarstackComponent.prototype.drillid;
    /** @type {?} */
    BarstackComponent.prototype.onLegendClick;
    /** @type {?} */
    BarstackComponent.prototype.httpresponse;
    /** @type {?} */
    BarstackComponent.prototype.svg;
    /** @type {?} */
    BarstackComponent.prototype.offsetheight;
    /** @type {?} */
    BarstackComponent.prototype.myservice;
    /** @type {?} */
    BarstackComponent.prototype.device;
}
