// amexio-chart-d3/esm5/lib/groupbar/groupbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, Output, EventEmitter } from '@angular/core';
import { AmexioD3BaseChartComponent } from "../base/base.component";
import { CommanDataService } from '../services/comman.data.service';
import { DeviceQueryService } from '../services/device.query.service';
import * as d3 from 'd3';
var GroupbarComponent = /** @class */ (function (_super) {
    tslib_1.__extends(GroupbarComponent, _super);
    function GroupbarComponent(myservice, device) {
        var _this = _super.call(this, 'multibar') || this;
        _this.myservice = myservice;
        _this.device = device;
        _this.legend = true;
        _this.barwidth = 0;
        _this.labelcolor = "black";
        _this.labelflag = false;
        _this.onLegendClick = new EventEmitter();
        _this.svgwidth = 300;
        _this.svgheight = 300;
        _this.level = 0;
        _this.drillabledatakey = [];
        _this.hScale = true;
        _this.drillableFlag = true;
        _this.resizeflag = false;
        _this.groupbarchartArray = [];
        _this.urllegendArray = [];
        _this.predefinedcolors = ["#3366cc", "#dc3912", "#ff9900", "#109618", "#990099", "#0099c6", "#dd4477", "#66aa00", "#b82e2e", "#316395", "#994499", "#22aa99", "#aaaa11", "#6633cc", "#e67300", "#8b0707", "#651067", "#329262", "#5574a6", "#3b3eac"];
        return _this;
    }
    /**
     * @return {?}
     */
    GroupbarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.level <= 1) {
            /** @type {?} */
            var res_1;
            if (this.httpmethod && this.httpurl) {
                this.myservice.fetchUrlData(this.httpurl, this.httpmethod).subscribe(function (response) {
                    res_1 = response;
                }, function (error) {
                }, function () {
                    setTimeout(function () {
                        _this.data = _this.getResponseData(res_1);
                        _this.initializeData();
                        _this.plotD3Chart();
                    }, 0);
                });
            }
            else if (this.data) {
                setTimeout(function () {
                    _this.data = _this.getResponseData(_this.data);
                    _this.initializeData();
                    _this.plotD3Chart();
                }, 0);
            }
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    GroupbarComponent.prototype.fetchData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        /** @type {?} */
        var requestJson;
        /** @type {?} */
        var key = this.drillabledatakey;
        /** @type {?} */
        var resp;
        if (this.drillabledatakey.length) {
            /** @type {?} */
            var drillabledata = this.getMultipleDrillbleKeyData(data, key);
            requestJson = drillabledata;
        }
        else {
            requestJson = data;
        }
        if (this.httpmethod && this.httpurl) {
            this.myservice.postfetchData(this.httpurl, this.httpmethod, requestJson).subscribe(function (response) {
                resp = response;
            }, function (error) {
            }, function () {
                setTimeout(function () {
                    _this.data = _this.getResponseData(resp);
                    _this.drawChart();
                }, 0);
            });
        }
    };
    /**
     * @return {?}
     */
    GroupbarComponent.prototype.drawChart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () {
            _this.initializeData();
            _this.plotD3Chart();
        }, 0);
    };
    /**
     * @param {?} httpResponse
     * @return {?}
     */
    GroupbarComponent.prototype.getResponseData = /**
     * @param {?} httpResponse
     * @return {?}
     */
    function (httpResponse) {
        /** @type {?} */
        var responsedata = httpResponse;
        if (this.datareader != null) {
            /** @type {?} */
            var dr = this.datareader.split('.');
            try {
                for (var dr_1 = tslib_1.__values(dr), dr_1_1 = dr_1.next(); !dr_1_1.done; dr_1_1 = dr_1.next()) {
                    var ir = dr_1_1.value;
                    responsedata = responsedata[ir];
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (dr_1_1 && !dr_1_1.done && (_a = dr_1.return)) _a.call(dr_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        else {
            responsedata = httpResponse;
        }
        return responsedata;
        var e_1, _a;
    };
    /**
     * @return {?}
     */
    GroupbarComponent.prototype.plotD3Chart = /**
     * @return {?}
     */
    function () {
        this.convertToJSON();
        this.plotGroupBarChart();
        this.transformData(this.data);
    };
    /**
     * @return {?}
     */
    GroupbarComponent.prototype.plotGroupBarChart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var tooltip = this.toolTip(d3);
        /** @type {?} */
        var colors = this.predefinedcolors;
        // this.svgwidth = this.chartId.nativeElement.offsetWidth;
        if (this.resizeflag == false) {
            if (this.chartId) {
                this.svgwidth = this.chartId.nativeElement.offsetWidth;
            }
            else {
                this.svgwidth = this.svgwidth;
            }
        }
        /** @type {?} */
        var margin = { top: 20, right: 20, bottom: 50, left: 40 };
        /** @type {?} */
        var width = this.svgwidth - margin.left - margin.right;
        /** @type {?} */
        var height = this.svgheight - margin.top - margin.bottom;
        this.svg = d3.select("#" + this.componentId)
            .attr("width", width + margin.left + margin.right)
            .attr("height", height + margin.top + margin.bottom)
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
        /** @type {?} */
        var x0 = d3.scaleBand()
            .rangeRound([0, width])
            .padding(0.2);
        /** @type {?} */
        var x1 = d3.scaleBand().padding(0.1);
        /** @type {?} */
        var y = d3.scaleLinear()
            .rangeRound([height, 0]);
        //setting x and y domains
        this.years = this.groupbarchartArray.map(function (d) { return d.labels; });
        /** @type {?} */
        var label = this.groupbarchartArray[0].values.map(function (d) { return d.label; });
        x0.domain(this.years);
        x1.domain(label).rangeRound([0, x0.bandwidth()]);
        y.domain([0, d3.max(this.groupbarchartArray, function (labels) { return d3.max(labels.values, function (d) { return d.value; }); })]);
        //dynamic barwidth
        if (this.barwidth > 0) {
            this.barwidth = this.barwidth;
        }
        else {
            this.barwidth = x0.bandwidth;
        }
        // add x axis to svg
        if (this.device.IsDesktop() == true) {
            this.svg.append("g")
                .attr("transform", "translate(0," + height + ")")
                .call(d3.axisBottom(x0));
        }
        else {
            this.svg.append("g")
                .attr("transform", "translate(0," + height + ")")
                .call(d3.axisBottom(x0)).
                selectAll("text")
                .attr("y", 0)
                .attr("x", 9)
                .attr("dy", ".35em")
                .attr("transform", "rotate(60)")
                .style("text-anchor", "start");
        }
        //add y axis to svg
        this.svg.append("g")
            .call(d3.axisLeft(y)
            .ticks(10));
        this.plotLine(this.svg, y, height, width);
        // svg.select('.y').transition().duration(500).delay(1300).style('opacity', '1');
        //adding bars
        /** @type {?} */
        var slice = this.svg.selectAll(".slice")
            .data(this.groupbarchartArray)
            .enter().append("g")
            .attr("class", "g")
            .attr("transform", function (d) { return "translate(" + x0(d.labels) + ",0)"; });
        slice.selectAll("rect")
            .data(function (d) { return d.values; })
            .enter().append("rect")
            .attr("width", x1.bandwidth)
            .attr("x", function (d) {
            return x1(d.label);
        })
            .style("fill", function (d, index) { return colors[index]; })
            .attr("y", function (d) { return y(0); })
            .attr("height", function (d) { return height - y(0); })
            .attr("cursor", "pointer")
            .on("mouseover", function (d) {
            return tooltip.style("visibility", "visible");
        })
            .on("mousemove", function (d) {
            return tooltip.html(_this.setKey(d)
            //  this.toolTipContent(d)
            )
                .style("top", (d3.event.pageY - 10) + "px")
                .style("left", (d3.event.pageX + 10) + "px");
        }).on("mouseout", function (d) {
            return tooltip.style("visibility", "hidden");
        })
            .on("click", function (d) {
            _this.groupbarClick(d);
            _this.fordrillableClick(_this, d, event);
            return tooltip.style("visibility", "hidden");
            // this.chartClick(d);
        });
        // -------------------------------------------------------
        if (this.labelflag) {
            slice.selectAll("text")
                .data(function (d) { return d.values; })
                .enter().append("text")
                .attr("width", x1.bandwidth)
                .attr("x", function (d) {
                return x1(d.label) + x1.bandwidth() / 2;
            })
                .attr("y", function (d) { return y(d.value); })
                .attr("height", function (d) { return height - y(0); })
                .style("font-weight", "bold")
                .style("font-size", "1vw")
                .attr("text-anchor", "middle")
                .attr("fill", function (d) {
                if (_this.labelcolor.length > 0) {
                    return _this.labelcolor;
                }
                else {
                    return "black";
                }
            })
                .text(function (d) {
                return d.value;
            });
        }
        slice.selectAll("rect")
            .attr("y", function (d) {
            return y(d.value);
        })
            .attr("height", function (d) { return height - y(d.value); });
    };
    /**
     * @param {?} d
     * @return {?}
     */
    GroupbarComponent.prototype.groupbarClick = /**
     * @param {?} d
     * @return {?}
     */
    function (d) {
        /** @type {?} */
        var object = {};
        object[d.label] = d.value;
        object[this.xaxisData] = d.xaxis;
        this.chartClick(object);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    GroupbarComponent.prototype.resize = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.svgwidth = 0;
        this.svg.selectAll("*").remove();
        this.resizeflag = true;
        this.svgwidth = this.divid.nativeElement.offsetWidth;
        this.plotGroupBarChart();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    GroupbarComponent.prototype.legendClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // const legendNode = JSON.parse(JSON.stringify(event));
        // delete legendNode.color;
        // const legendNode = JSON.parse(JSON.stringify(event));
        // delete legendNode.color;
        /** @type {?} */
        var obj = {};
        obj["label"] = event.label;
        /** @type {?} */
        var data = [];
        event.data.forEach(function (element) {
            /** @type {?} */
            var object = {};
            object[element.label] = element.value;
            data.push(object);
        });
        obj["data"] = data;
        this.onLegendClick.emit(obj);
        // this.onLegendClick.emit(legendNode);
    };
    //2d array to json conversion
    //2d array to json conversion
    /**
     * @return {?}
     */
    GroupbarComponent.prototype.convertToJSON = 
    //2d array to json conversion
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var groupChartObj = { "labels": "", values: [] };
        /** @type {?} */
        var firstRowOfData = this.data[0];
        this.xaxisData = this.data[0][0];
        for (var i = 1; i < this.data.length; i++) {
            /** @type {?} */
            var multiSeriesArray = [];
            /** @type {?} */
            var valueOfJ = void 0;
            for (var j = 1; j < this.data[i].length; j++) {
                valueOfJ = this.data[i][0];
                /** @type {?} */
                var singleBarObj = {};
                singleBarObj["value"] = this.data[i][j];
                singleBarObj["label"] = firstRowOfData[j];
                singleBarObj["xaxis"] = this.data[i][0];
                multiSeriesArray.push(singleBarObj);
            }
            if (multiSeriesArray.length) {
                groupChartObj["values"] = multiSeriesArray;
                /** @type {?} */
                var newLabelsValues = valueOfJ;
                /** @type {?} */
                var newGroupDataObj = Object.assign({}, groupChartObj);
                newGroupDataObj['labels'] = newLabelsValues + '';
                this.groupbarchartArray.push(newGroupDataObj);
            }
        }
    };
    /**
     * @return {?}
     */
    GroupbarComponent.prototype.transformto2dArray = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var i = 0;
        /** @type {?} */
        var result = [];
        result.push("years");
        this.data.forEach(function (element, i) {
            if (i == 0) {
                element.values.forEach(function (element2) {
                    result.push(element2.label);
                });
            }
        });
        this.urllegendArray.push(result);
        /** @type {?} */
        var temparray = [];
        this.data.forEach(function (element) {
            temparray = [];
            /** @type {?} */
            var values;
            /** @type {?} */
            var year = element.labels;
            temparray.push(year);
            element.values.forEach(function (element2) {
                values = element2.value;
                temparray.push(values);
            });
            _this.urllegendArray.push(temparray);
        });
    };
    /**
     * @param {?} data
     * @return {?}
     */
    GroupbarComponent.prototype.transformData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.keyArray = [];
        this.legendArray = [];
        data.forEach(function (element, i) {
            if (i == 0) {
                element.forEach(function (innerelement, index) {
                    if (index > 0) {
                        _this.legendArray[innerelement] = { 'data': [] };
                        _this.keyArray.push(innerelement);
                    }
                });
            }
        });
        data.forEach(function (element, index) {
            if (index > 0) {
                /** @type {?} */
                var obj_1 = {};
                element.forEach(function (innerelement, innerindex) {
                    if (innerindex >= 0) {
                        /** @type {?} */
                        var key = _this.keyArray[innerindex - 1];
                        obj_1[key] = element[innerindex];
                        /** @type {?} */
                        var legenddata = _this.legendArray[key];
                        if (legenddata) {
                            legenddata.data.push({ 'value': element[innerindex], 'label': element[0] });
                        }
                    }
                });
            }
        });
        this.legends = [];
        this.keyArray.forEach(function (element, index) {
            /** @type {?} */
            var legenddata = _this.legendArray[element];
            /** @type {?} */
            var object = { 'label': element, 'color': _this.predefinedcolors[index], 'data': legenddata.data };
            _this.legends.push(object);
        });
    };
    /**
     * @param {?} d
     * @return {?}
     */
    GroupbarComponent.prototype.setKey = /**
     * @param {?} d
     * @return {?}
     */
    function (d) {
        /** @type {?} */
        var object = {};
        object[d.label] = d.value;
        object[this.xaxisData] = d.xaxis;
        return (this.toolTipForBar(object));
    };
    /**
     * @param {?} g
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    GroupbarComponent.prototype.plotLine = /**
     * @param {?} g
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    function (g, y, height, width) {
        if (this.hScale) {
            g.append('g')
                .attr("color", "lightgrey")
                .call(d3.axisLeft(y)
                .tickSize(-width).tickFormat(''));
        }
    };
    GroupbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'amexio-d3-chart-multiseries',
                    template: "<div   *ngIf=\"drillableFlag\" #divid (window:resize)=\"resize($event)\"> \n<table style=\"width: 100%;\" #chartId>\n    <tr>\n        <td colspan=\"2\" style=\"font-size: 18px; font-weight: bold; \"> {{title}} </td>\n    </tr>\n    <tr>\n        <td>\n            <svg [attr.height]=\"height\" [attr.width]=\"width\" [attr.id]=\"componentId\">\n            </svg>\n        </td>\n    </tr>\n    <tr>\n        <td style=\"display: block; padding: 5px;\">\n            <amexio-d3-legend *ngIf=\"legend\" [horizontal]=\"true\" (onClick)=\"legendClick($event)\" [data]=\"legends\" style=\"height:100%;\"></amexio-d3-legend>\n        </td>\n    </tr>\n</table>\n</div>\n\n\n",
                    styles: [""]
                },] },
    ];
    GroupbarComponent.ctorParameters = function () { return [
        { type: CommanDataService },
        { type: DeviceQueryService }
    ]; };
    GroupbarComponent.propDecorators = {
        chartId: [{ type: ViewChild, args: ['chartId',] }],
        divid: [{ type: ViewChild, args: ['divid',] }],
        drillid: [{ type: ViewChild, args: ['drillid',] }],
        data: [{ type: Input, args: ['data',] }],
        legend: [{ type: Input }],
        barwidth: [{ type: Input }],
        labelcolor: [{ type: Input, args: ['label-color',] }],
        labelflag: [{ type: Input, args: ['label',] }],
        onLegendClick: [{ type: Output }],
        svgwidth: [{ type: Input, args: ['width',] }],
        svgheight: [{ type: Input, args: ['height',] }],
        datareader: [{ type: Input, args: ['data-reader',] }],
        level: [{ type: Input, args: ['level',] }],
        target: [{ type: Input, args: ['target',] }],
        drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
        hScale: [{ type: Input, args: ['horizontal-scale',] }]
    };
    return GroupbarComponent;
}(AmexioD3BaseChartComponent));
export { GroupbarComponent };
if (false) {
    /** @type {?} */
    GroupbarComponent.prototype.chartId;
    /** @type {?} */
    GroupbarComponent.prototype.divid;
    /** @type {?} */
    GroupbarComponent.prototype.drillid;
    /** @type {?} */
    GroupbarComponent.prototype.data;
    /** @type {?} */
    GroupbarComponent.prototype.legend;
    /** @type {?} */
    GroupbarComponent.prototype.barwidth;
    /** @type {?} */
    GroupbarComponent.prototype.labelcolor;
    /** @type {?} */
    GroupbarComponent.prototype.labelflag;
    /** @type {?} */
    GroupbarComponent.prototype.onLegendClick;
    /** @type {?} */
    GroupbarComponent.prototype.svgwidth;
    /** @type {?} */
    GroupbarComponent.prototype.svgheight;
    /** @type {?} */
    GroupbarComponent.prototype.datareader;
    /** @type {?} */
    GroupbarComponent.prototype.level;
    /** @type {?} */
    GroupbarComponent.prototype.target;
    /** @type {?} */
    GroupbarComponent.prototype.drillabledatakey;
    /** @type {?} */
    GroupbarComponent.prototype.hScale;
    /** @type {?} */
    GroupbarComponent.prototype.drillableFlag;
    /** @type {?} */
    GroupbarComponent.prototype.resizeflag;
    /** @type {?} */
    GroupbarComponent.prototype.groupbarchartArray;
    /** @type {?} */
    GroupbarComponent.prototype.legendArray;
    /** @type {?} */
    GroupbarComponent.prototype.xaxisData;
    /** @type {?} */
    GroupbarComponent.prototype.keyArray;
    /** @type {?} */
    GroupbarComponent.prototype.legends;
    /** @type {?} */
    GroupbarComponent.prototype.years;
    /** @type {?} */
    GroupbarComponent.prototype.urllegendArray;
    /** @type {?} */
    GroupbarComponent.prototype.svg;
    /** @type {?} */
    GroupbarComponent.prototype.myservice;
    /** @type {?} */
    GroupbarComponent.prototype.device;
}
