// amexio-chart-d3/esm2015/lib/candlestick/candlestick.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef, Output, EventEmitter, } from "@angular/core";
import { AmexioD3BaseChartComponent } from "../base/base.component";
import { CommanDataService } from '../services/comman.data.service';
import { DeviceQueryService } from '../services/device.query.service';
import * as d3 from 'd3';
export class CandlestickComponent extends AmexioD3BaseChartComponent {
    /**
     * @param {?} myservice
     * @param {?} device
     */
    constructor(myservice, device) {
        super("candlestickwaterfallchart");
        this.myservice = myservice;
        this.device = device;
        this.svgwidth = 300;
        this.svgheight = 300;
        this.level = 0;
        this.drillabledatakey = [];
        this.labelcolor = "black";
        this.labelflag = false;
        this.hScale = true;
        this.onLegendClick = new EventEmitter();
        this.drillableFlag = true;
        this.resizeflag = false;
        this.predefinedColor = [];
        this.keyArray = [];
        this.margin = {};
        this.legendArray = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.predefinedColor = ["#3366cc", "#dc3912", "#ff9900", "#109618", "#990099", "#0099c6", "#dd4477", "#66aa00", "#b82e2e", "#316395", "#994499", "#22aa99", "#aaaa11", "#6633cc", "#e67300", "#8b0707", "#651067", "#329262", "#5574a6", "#3b3eac"];
        if (this.level <= 1) {
            /** @type {?} */
            let res;
            if (this.httpmethod && this.httpurl) {
                this.myservice.fetchUrlData(this.httpurl, this.httpmethod).subscribe((response) => {
                    //this.data = response;
                    this.httpresponse = response;
                    this.data = this.getResponseData(response);
                }, (error) => {
                }, () => {
                    setTimeout(() => {
                        this.transformData(this.data);
                        this.initializeData();
                        this.plotXYAxis();
                        this.plotD3Chart();
                    }, 0);
                });
            }
            else if (this.data) {
                setTimeout(() => {
                    this.transformData(this.data);
                    this.initializeData();
                    this.plotXYAxis();
                    this.plotD3Chart();
                }, 0);
            }
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    fetchData(data) {
        /** @type {?} */
        let requestJson;
        /** @type {?} */
        let key = this.drillabledatakey;
        /** @type {?} */
        let resp;
        if (this.drillabledatakey.length) {
            /** @type {?} */
            let drillabledata = this.getMultipleDrillbleKeyData(data, key);
            requestJson = drillabledata;
        }
        else {
            requestJson = data;
        }
        if (this.httpmethod && this.httpurl) {
            this.myservice.postfetchData(this.httpurl, this.httpmethod, requestJson).subscribe((response) => {
                resp = response;
                this.httpresponse = response;
            }, (error) => {
            }, () => {
                setTimeout(() => {
                    //this.data = this.getResponseData(resp);
                    this.drawChart();
                }, 0);
            });
        }
    }
    /**
     * @return {?}
     */
    drawChart() {
        setTimeout(() => {
            this.data = this.getResponseData(this.httpresponse);
            this.transformData(this.data);
            this.initializeData();
            this.plotXYAxis();
            this.plotD3Chart();
        }, 0);
    }
    /**
     * @return {?}
     */
    initializeData() {
        this.tooltip = this.toolTip(d3);
        if (this.resizeflag == false) {
            if (this.chartId) {
                this.svgwidth = this.chartId.nativeElement.offsetWidth;
            }
            else {
                this.svgwidth = this.svgwidth;
            }
        }
        this.margin = { top: 20, right: 30, bottom: 50, left: 60 },
            this.width = this.svgwidth - this.margin.left - this.margin.right,
            this.height = this.svgheight - this.margin.top - this.margin.bottom;
    }
    /**
     * @return {?}
     */
    plotXYAxis() {
        // set the ranges
        this.x = d3.scaleBand().range([0, this.width]);
        this.y = d3.scaleLinear()
            .rangeRound([this.height, 0]);
        // scale the range of the data
        /** @type {?} */
        let candlestickArray = this.data.map((d) => {
            return d[Object.keys(d)[0]];
        });
        this.x.domain(candlestickArray);
        /** @type {?} */
        let max = d3.max(this.data, (d) => { return d.end; });
        this.y.domain([0, max]);
        this.svg = d3.select("#" + this.componentId)
            .attr("width", this.width + this.margin.left + this.margin.right)
            .attr("height", this.height + this.margin.top + this.margin.bottom)
            .append("g")
            .attr("transform", "translate(" + this.margin.left + "," + this.margin.top + ")");
        // add the X Axis
        if (this.device.IsDesktop() == true) {
            this.svg.append("g")
                .attr("transform", "translate(0," + this.height + ")")
                .call(d3.axisBottom(this.x));
        }
        else {
            this.svg.append("g")
                .attr("transform", "translate(0," + this.height + ")")
                .call(d3.axisBottom(this.x)).
                selectAll("text")
                .attr("y", 0)
                .attr("x", 9)
                .attr("dy", ".35em")
                .attr("transform", "rotate(60)")
                .style("text-anchor", "start");
        }
        // this.svg.append("g")
        //   .attr("transform", "translate(0," + this.height + ")")
        //   .call(d3.axisBottom(this.x));
        // add the Y Axis
        this.svg.append("g")
            .call(d3.axisLeft(this.y));
        this.plotLine(this.svg, this.x, this.y, this.height, this.width);
    }
    /**
     * @return {?}
     */
    plotD3Chart() {
        /** @type {?} */
        let bar = this.svg.selectAll(".bar")
            .data(this.data)
            .enter().append("g")
            .attr("class", (d) => { return "bar " + d.class; })
            .attr("transform", (d) => {
            return "translate(" + this.x(d[Object.keys(d)[0]]) + ",0)";
        });
        bar.append("rect")
            .attr("y", (d) => { return this.y(Math.max(d.start, d.end)); })
            .attr("height", (d) => { return Math.abs(this.y(d.start) - this.y(d.end)); })
            .attr("width", this.x.bandwidth())
            .attr("fill", (d, i) => {
            return this.predefinedColor[i];
        })
            .attr("cursor", "pointer")
            .on("mouseover", (d) => {
            this.formTooltipData(d);
            return this.tooltip.style("visibility", "visible");
        })
            .on("mousemove", (d) => {
            return this.tooltip.html(this.formTooltipData(d))
                .style("top", (d3.event.pageY - 10) + "px")
                .style("left", (d3.event.pageX + 10) + "px");
        })
            .on("mouseout", (d) => {
            return this.tooltip.style("visibility", "hidden");
        })
            .on("click", (d) => {
            this.onCandlestickClick(d);
            this.fordrillableClick(this, d, event);
            return this.tooltip.style("visibility", "hidden");
        });
        if (this.labelflag) {
            bar.append("text")
                .style("font-weight", "bold")
                .style("font-size", "1vw")
                .attr("text-anchor", "middle")
                .attr("fill", (d) => {
                if (this.labelcolor.length > 0) {
                    return this.labelcolor;
                }
                else {
                    return "black";
                }
            })
                .attr("x", (d) => {
                return (this.x.bandwidth()) / 2;
            })
                .attr("y", (d, index) => {
                return this.y(Math.max(d.start, d.end)) + 20;
            })
                .text(function (d) {
                return d[Object.keys(d)[1]];
            });
        }
    }
    /**
     * @param {?} svg
     * @param {?} x
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    plotLine(svg, x, y, height, width) {
        if (this.hScale) {
            svg.append('g')
                .attr("color", "lightgrey")
                .call(d3.axisLeft(y)
                .tickSize(-width).tickFormat(''));
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    transformData(data) {
        this.transformeddata = [];
        this.keyArray = [];
        this.keyArray = data[0];
        data.forEach((element, index) => {
            if (index > 0) {
                /** @type {?} */
                let DummyObject = {};
                element.forEach((individualvalue, keyindex) => {
                    DummyObject[this.keyArray[keyindex]] = individualvalue;
                }); //inner for loop ends
                this.transformeddata.push(DummyObject);
            } //if ends
        }); //outer for loop ends 
        this.data = this.transformeddata;
        this.addDataKeys();
        this.formLegendData();
    }
    /**
     * @return {?}
     */
    addDataKeys() {
        /** @type {?} */
        let cumulative = 0;
        for (let i = 0; i < this.data.length; i++) {
            this.data[i]["start"] = cumulative;
            cumulative += this.data[i][this.keyArray[1]];
            this.data[i]["end"] = cumulative;
            this.data[i]["class"] = (this.data[i][this.keyArray[1]] >= 0) ? 'positive' : 'negative';
        }
    }
    /**
     * @param {?} httpResponse
     * @return {?}
     */
    getResponseData(httpResponse) {
        /** @type {?} */
        let responsedata = httpResponse;
        if (this.datareader != null) {
            /** @type {?} */
            const dr = this.datareader.split('.');
            for (const ir of dr) {
                responsedata = responsedata[ir];
            }
        }
        else {
            responsedata = httpResponse;
        }
        return responsedata;
    }
    /**
     * @param {?} tooltipData
     * @return {?}
     */
    formTooltipData(tooltipData) {
        /** @type {?} */
        let object = {};
        object[this.keyArray[0]] = tooltipData[Object.keys(tooltipData)[0]];
        object[this.keyArray[1]] = tooltipData[Object.keys(tooltipData)[1]];
        return this.toolTipForBar(object);
    }
    /**
     * @param {?} chartData
     * @return {?}
     */
    onCandlestickClick(chartData) {
        /** @type {?} */
        let object = {};
        object[this.keyArray[0]] = chartData[Object.keys(chartData)[0]];
        object[this.keyArray[1]] = chartData[Object.keys(chartData)[1]];
        this.chartClick(object);
    }
    /**
     * @return {?}
     */
    formLegendData() {
        this.legendArray = [];
        this.data.forEach((element, index) => {
            for (let [key, value] of Object.entries(element)) {
                if (key == this.keyArray[0]) {
                    /** @type {?} */
                    let object = {};
                    object["label"] = value;
                    object["color"] = this.predefinedColor[index];
                    this.legendArray.push(object);
                }
            }
        });
    }
    /**
     * @param {?} chartData
     * @return {?}
     */
    onCandlestickLegendClick(chartData) {
        /** @type {?} */
        let object = {};
        this.data.forEach(element => {
            for (let [key, value] of Object.entries(element)) {
                if (value == chartData.label) {
                    object[chartData.label] = element.value;
                }
            }
        });
        this.onLegendClick.emit(object);
    }
    /**
     * @return {?}
     */
    resize() {
        this.svgwidth = 0;
        this.svg.selectAll("*").remove();
        this.resizeflag = true;
        this.svgwidth = this.divid.nativeElement.offsetWidth;
        this.initializeData();
        this.plotXYAxis();
        this.plotD3Chart();
    }
}
CandlestickComponent.decorators = [
    { type: Component, args: [{
                selector: 'amexio-d3-chart-waterfall',
                template: `
<div   *ngIf="drillableFlag" #divid (window:resize)="resize()" >
<table style="width: 100%;" #chartId (window:resize)="resize()">
  <tr>
      <td style="font-size: 18px; font-weight: bold; "> {{title}} </td>
  </tr>
  <tr>
      <td>
          <svg [attr.height]="svgheight" [attr.width]="svgwidth" [attr.id]="componentId">
          </svg>
      </td>
  </tr>
  <tr>
      <td style="display: block; padding: 5px;">
          <amexio-d3-legend *ngIf="legend" [horizontal]="true" (onClick)="onCandlestickLegendClick($event)" [data]="legendArray" style="height:100%;">
          </amexio-d3-legend>
      </td>
  </tr>
</table>
</div>`,
                styles: [``]
            },] },
];
CandlestickComponent.ctorParameters = () => [
    { type: CommanDataService },
    { type: DeviceQueryService }
];
CandlestickComponent.propDecorators = {
    svgwidth: [{ type: Input, args: ['width',] }],
    svgheight: [{ type: Input, args: ['height',] }],
    datareader: [{ type: Input, args: ['data-reader',] }],
    data: [{ type: Input }],
    level: [{ type: Input, args: ['level',] }],
    target: [{ type: Input, args: ['target',] }],
    drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
    labelcolor: [{ type: Input, args: ['label-color',] }],
    labelflag: [{ type: Input, args: ['label',] }],
    hScale: [{ type: Input, args: ['horizontal-scale',] }],
    chartId: [{ type: ViewChild, args: ['chartId',] }],
    divid: [{ type: ViewChild, args: ['divid',] }],
    onLegendClick: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CandlestickComponent.prototype.svgwidth;
    /** @type {?} */
    CandlestickComponent.prototype.svgheight;
    /** @type {?} */
    CandlestickComponent.prototype.datareader;
    /** @type {?} */
    CandlestickComponent.prototype.data;
    /** @type {?} */
    CandlestickComponent.prototype.level;
    /** @type {?} */
    CandlestickComponent.prototype.target;
    /** @type {?} */
    CandlestickComponent.prototype.drillabledatakey;
    /** @type {?} */
    CandlestickComponent.prototype.labelcolor;
    /** @type {?} */
    CandlestickComponent.prototype.labelflag;
    /** @type {?} */
    CandlestickComponent.prototype.hScale;
    /** @type {?} */
    CandlestickComponent.prototype.chartId;
    /** @type {?} */
    CandlestickComponent.prototype.divid;
    /** @type {?} */
    CandlestickComponent.prototype.onLegendClick;
    /** @type {?} */
    CandlestickComponent.prototype.drillableFlag;
    /** @type {?} */
    CandlestickComponent.prototype.resizeflag;
    /** @type {?} */
    CandlestickComponent.prototype.predefinedColor;
    /** @type {?} */
    CandlestickComponent.prototype.keyArray;
    /** @type {?} */
    CandlestickComponent.prototype.transformeddata;
    /** @type {?} */
    CandlestickComponent.prototype.height;
    /** @type {?} */
    CandlestickComponent.prototype.width;
    /** @type {?} */
    CandlestickComponent.prototype.margin;
    /** @type {?} */
    CandlestickComponent.prototype.x;
    /** @type {?} */
    CandlestickComponent.prototype.y;
    /** @type {?} */
    CandlestickComponent.prototype.svg;
    /** @type {?} */
    CandlestickComponent.prototype.tooltip;
    /** @type {?} */
    CandlestickComponent.prototype.legendArray;
    /** @type {?} */
    CandlestickComponent.prototype.httpresponse;
    /** @type {?} */
    CandlestickComponent.prototype.myservice;
    /** @type {?} */
    CandlestickComponent.prototype.device;
}
