// amexio-chart-d3/esm5/lib/candlestick/candlestick.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, ElementRef, Output, EventEmitter, } from "@angular/core";
import { AmexioD3BaseChartComponent } from "../base/base.component";
import { CommanDataService } from '../services/comman.data.service';
import { DeviceQueryService } from '../services/device.query.service';
import * as d3 from 'd3';
var CandlestickComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CandlestickComponent, _super);
    function CandlestickComponent(myservice, device) {
        var _this = _super.call(this, "candlestickwaterfallchart") || this;
        _this.myservice = myservice;
        _this.device = device;
        _this.svgwidth = 300;
        _this.svgheight = 300;
        _this.level = 0;
        _this.drillabledatakey = [];
        _this.labelcolor = "black";
        _this.labelflag = false;
        _this.hScale = true;
        _this.onLegendClick = new EventEmitter();
        _this.drillableFlag = true;
        _this.resizeflag = false;
        _this.predefinedColor = [];
        _this.keyArray = [];
        _this.margin = {};
        _this.legendArray = [];
        return _this;
    }
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.predefinedColor = ["#3366cc", "#dc3912", "#ff9900", "#109618", "#990099", "#0099c6", "#dd4477", "#66aa00", "#b82e2e", "#316395", "#994499", "#22aa99", "#aaaa11", "#6633cc", "#e67300", "#8b0707", "#651067", "#329262", "#5574a6", "#3b3eac"];
        if (this.level <= 1) {
            /** @type {?} */
            var res = void 0;
            if (this.httpmethod && this.httpurl) {
                this.myservice.fetchUrlData(this.httpurl, this.httpmethod).subscribe(function (response) {
                    //this.data = response;
                    _this.httpresponse = response;
                    _this.data = _this.getResponseData(response);
                }, function (error) {
                }, function () {
                    setTimeout(function () {
                        _this.transformData(_this.data);
                        _this.initializeData();
                        _this.plotXYAxis();
                        _this.plotD3Chart();
                    }, 0);
                });
            }
            else if (this.data) {
                setTimeout(function () {
                    _this.transformData(_this.data);
                    _this.initializeData();
                    _this.plotXYAxis();
                    _this.plotD3Chart();
                }, 0);
            }
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CandlestickComponent.prototype.fetchData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        /** @type {?} */
        var requestJson;
        /** @type {?} */
        var key = this.drillabledatakey;
        /** @type {?} */
        var resp;
        if (this.drillabledatakey.length) {
            /** @type {?} */
            var drillabledata = this.getMultipleDrillbleKeyData(data, key);
            requestJson = drillabledata;
        }
        else {
            requestJson = data;
        }
        if (this.httpmethod && this.httpurl) {
            this.myservice.postfetchData(this.httpurl, this.httpmethod, requestJson).subscribe(function (response) {
                resp = response;
                _this.httpresponse = response;
            }, function (error) {
            }, function () {
                setTimeout(function () {
                    //this.data = this.getResponseData(resp);
                    _this.drawChart();
                }, 0);
            });
        }
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.drawChart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () {
            _this.data = _this.getResponseData(_this.httpresponse);
            _this.transformData(_this.data);
            _this.initializeData();
            _this.plotXYAxis();
            _this.plotD3Chart();
        }, 0);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.initializeData = /**
     * @return {?}
     */
    function () {
        this.tooltip = this.toolTip(d3);
        if (this.resizeflag == false) {
            if (this.chartId) {
                this.svgwidth = this.chartId.nativeElement.offsetWidth;
            }
            else {
                this.svgwidth = this.svgwidth;
            }
        }
        this.margin = { top: 20, right: 30, bottom: 50, left: 60 },
            this.width = this.svgwidth - this.margin.left - this.margin.right,
            this.height = this.svgheight - this.margin.top - this.margin.bottom;
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.plotXYAxis = /**
     * @return {?}
     */
    function () {
        // set the ranges
        this.x = d3.scaleBand().range([0, this.width]);
        this.y = d3.scaleLinear()
            .rangeRound([this.height, 0]);
        // scale the range of the data
        /** @type {?} */
        var candlestickArray = this.data.map(function (d) {
            return d[Object.keys(d)[0]];
        });
        this.x.domain(candlestickArray);
        /** @type {?} */
        var max = d3.max(this.data, function (d) { return d.end; });
        this.y.domain([0, max]);
        this.svg = d3.select("#" + this.componentId)
            .attr("width", this.width + this.margin.left + this.margin.right)
            .attr("height", this.height + this.margin.top + this.margin.bottom)
            .append("g")
            .attr("transform", "translate(" + this.margin.left + "," + this.margin.top + ")");
        // add the X Axis
        if (this.device.IsDesktop() == true) {
            this.svg.append("g")
                .attr("transform", "translate(0," + this.height + ")")
                .call(d3.axisBottom(this.x));
        }
        else {
            this.svg.append("g")
                .attr("transform", "translate(0," + this.height + ")")
                .call(d3.axisBottom(this.x)).
                selectAll("text")
                .attr("y", 0)
                .attr("x", 9)
                .attr("dy", ".35em")
                .attr("transform", "rotate(60)")
                .style("text-anchor", "start");
        }
        // this.svg.append("g")
        //   .attr("transform", "translate(0," + this.height + ")")
        //   .call(d3.axisBottom(this.x));
        // add the Y Axis
        this.svg.append("g")
            .call(d3.axisLeft(this.y));
        this.plotLine(this.svg, this.x, this.y, this.height, this.width);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.plotD3Chart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var bar = this.svg.selectAll(".bar")
            .data(this.data)
            .enter().append("g")
            .attr("class", function (d) { return "bar " + d.class; })
            .attr("transform", function (d) {
            return "translate(" + _this.x(d[Object.keys(d)[0]]) + ",0)";
        });
        bar.append("rect")
            .attr("y", function (d) { return _this.y(Math.max(d.start, d.end)); })
            .attr("height", function (d) { return Math.abs(_this.y(d.start) - _this.y(d.end)); })
            .attr("width", this.x.bandwidth())
            .attr("fill", function (d, i) {
            return _this.predefinedColor[i];
        })
            .attr("cursor", "pointer")
            .on("mouseover", function (d) {
            _this.formTooltipData(d);
            return _this.tooltip.style("visibility", "visible");
        })
            .on("mousemove", function (d) {
            return _this.tooltip.html(_this.formTooltipData(d))
                .style("top", (d3.event.pageY - 10) + "px")
                .style("left", (d3.event.pageX + 10) + "px");
        })
            .on("mouseout", function (d) {
            return _this.tooltip.style("visibility", "hidden");
        })
            .on("click", function (d) {
            _this.onCandlestickClick(d);
            _this.fordrillableClick(_this, d, event);
            return _this.tooltip.style("visibility", "hidden");
        });
        if (this.labelflag) {
            bar.append("text")
                .style("font-weight", "bold")
                .style("font-size", "1vw")
                .attr("text-anchor", "middle")
                .attr("fill", function (d) {
                if (_this.labelcolor.length > 0) {
                    return _this.labelcolor;
                }
                else {
                    return "black";
                }
            })
                .attr("x", function (d) {
                return (_this.x.bandwidth()) / 2;
            })
                .attr("y", function (d, index) {
                return _this.y(Math.max(d.start, d.end)) + 20;
            })
                .text(function (d) {
                return d[Object.keys(d)[1]];
            });
        }
    };
    /**
     * @param {?} svg
     * @param {?} x
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    CandlestickComponent.prototype.plotLine = /**
     * @param {?} svg
     * @param {?} x
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    function (svg, x, y, height, width) {
        if (this.hScale) {
            svg.append('g')
                .attr("color", "lightgrey")
                .call(d3.axisLeft(y)
                .tickSize(-width).tickFormat(''));
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    CandlestickComponent.prototype.transformData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.transformeddata = [];
        this.keyArray = [];
        this.keyArray = data[0];
        data.forEach(function (element, index) {
            if (index > 0) {
                /** @type {?} */
                var DummyObject_1 = {};
                element.forEach(function (individualvalue, keyindex) {
                    DummyObject_1[_this.keyArray[keyindex]] = individualvalue;
                }); //inner for loop ends
                _this.transformeddata.push(DummyObject_1);
            } //if ends
        }); //outer for loop ends 
        this.data = this.transformeddata;
        this.addDataKeys();
        this.formLegendData();
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.addDataKeys = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var cumulative = 0;
        for (var i = 0; i < this.data.length; i++) {
            this.data[i]["start"] = cumulative;
            cumulative += this.data[i][this.keyArray[1]];
            this.data[i]["end"] = cumulative;
            this.data[i]["class"] = (this.data[i][this.keyArray[1]] >= 0) ? 'positive' : 'negative';
        }
    };
    /**
     * @param {?} httpResponse
     * @return {?}
     */
    CandlestickComponent.prototype.getResponseData = /**
     * @param {?} httpResponse
     * @return {?}
     */
    function (httpResponse) {
        /** @type {?} */
        var responsedata = httpResponse;
        if (this.datareader != null) {
            /** @type {?} */
            var dr = this.datareader.split('.');
            try {
                for (var dr_1 = tslib_1.__values(dr), dr_1_1 = dr_1.next(); !dr_1_1.done; dr_1_1 = dr_1.next()) {
                    var ir = dr_1_1.value;
                    responsedata = responsedata[ir];
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (dr_1_1 && !dr_1_1.done && (_a = dr_1.return)) _a.call(dr_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        else {
            responsedata = httpResponse;
        }
        return responsedata;
        var e_1, _a;
    };
    /**
     * @param {?} tooltipData
     * @return {?}
     */
    CandlestickComponent.prototype.formTooltipData = /**
     * @param {?} tooltipData
     * @return {?}
     */
    function (tooltipData) {
        /** @type {?} */
        var object = {};
        object[this.keyArray[0]] = tooltipData[Object.keys(tooltipData)[0]];
        object[this.keyArray[1]] = tooltipData[Object.keys(tooltipData)[1]];
        return this.toolTipForBar(object);
    };
    /**
     * @param {?} chartData
     * @return {?}
     */
    CandlestickComponent.prototype.onCandlestickClick = /**
     * @param {?} chartData
     * @return {?}
     */
    function (chartData) {
        /** @type {?} */
        var object = {};
        object[this.keyArray[0]] = chartData[Object.keys(chartData)[0]];
        object[this.keyArray[1]] = chartData[Object.keys(chartData)[1]];
        this.chartClick(object);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.formLegendData = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.legendArray = [];
        this.data.forEach(function (element, index) {
            try {
                for (var _a = tslib_1.__values(Object.entries(element)), _b = _a.next(); !_b.done; _b = _a.next()) {
                    var _c = tslib_1.__read(_b.value, 2), key = _c[0], value = _c[1];
                    if (key == _this.keyArray[0]) {
                        /** @type {?} */
                        var object = {};
                        object["label"] = value;
                        object["color"] = _this.predefinedColor[index];
                        _this.legendArray.push(object);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_d = _a.return)) _d.call(_a);
                }
                finally { if (e_2) throw e_2.error; }
            }
            var e_2, _d;
        });
    };
    /**
     * @param {?} chartData
     * @return {?}
     */
    CandlestickComponent.prototype.onCandlestickLegendClick = /**
     * @param {?} chartData
     * @return {?}
     */
    function (chartData) {
        /** @type {?} */
        var object = {};
        this.data.forEach(function (element) {
            try {
                for (var _a = tslib_1.__values(Object.entries(element)), _b = _a.next(); !_b.done; _b = _a.next()) {
                    var _c = tslib_1.__read(_b.value, 2), key = _c[0], value = _c[1];
                    if (value == chartData.label) {
                        object[chartData.label] = element.value;
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_b && !_b.done && (_d = _a.return)) _d.call(_a);
                }
                finally { if (e_3) throw e_3.error; }
            }
            var e_3, _d;
        });
        this.onLegendClick.emit(object);
    };
    /**
     * @return {?}
     */
    CandlestickComponent.prototype.resize = /**
     * @return {?}
     */
    function () {
        this.svgwidth = 0;
        this.svg.selectAll("*").remove();
        this.resizeflag = true;
        this.svgwidth = this.divid.nativeElement.offsetWidth;
        this.initializeData();
        this.plotXYAxis();
        this.plotD3Chart();
    };
    CandlestickComponent.decorators = [
        { type: Component, args: [{
                    selector: 'amexio-d3-chart-waterfall',
                    template: "\n<div   *ngIf=\"drillableFlag\" #divid (window:resize)=\"resize()\" >\n<table style=\"width: 100%;\" #chartId (window:resize)=\"resize()\">\n  <tr>\n      <td style=\"font-size: 18px; font-weight: bold; \"> {{title}} </td>\n  </tr>\n  <tr>\n      <td>\n          <svg [attr.height]=\"svgheight\" [attr.width]=\"svgwidth\" [attr.id]=\"componentId\">\n          </svg>\n      </td>\n  </tr>\n  <tr>\n      <td style=\"display: block; padding: 5px;\">\n          <amexio-d3-legend *ngIf=\"legend\" [horizontal]=\"true\" (onClick)=\"onCandlestickLegendClick($event)\" [data]=\"legendArray\" style=\"height:100%;\">\n          </amexio-d3-legend>\n      </td>\n  </tr>\n</table>\n</div>",
                    styles: [""]
                },] },
    ];
    CandlestickComponent.ctorParameters = function () { return [
        { type: CommanDataService },
        { type: DeviceQueryService }
    ]; };
    CandlestickComponent.propDecorators = {
        svgwidth: [{ type: Input, args: ['width',] }],
        svgheight: [{ type: Input, args: ['height',] }],
        datareader: [{ type: Input, args: ['data-reader',] }],
        data: [{ type: Input }],
        level: [{ type: Input, args: ['level',] }],
        target: [{ type: Input, args: ['target',] }],
        drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
        labelcolor: [{ type: Input, args: ['label-color',] }],
        labelflag: [{ type: Input, args: ['label',] }],
        hScale: [{ type: Input, args: ['horizontal-scale',] }],
        chartId: [{ type: ViewChild, args: ['chartId',] }],
        divid: [{ type: ViewChild, args: ['divid',] }],
        onLegendClick: [{ type: Output }]
    };
    return CandlestickComponent;
}(AmexioD3BaseChartComponent));
export { CandlestickComponent };
if (false) {
    /** @type {?} */
    CandlestickComponent.prototype.svgwidth;
    /** @type {?} */
    CandlestickComponent.prototype.svgheight;
    /** @type {?} */
    CandlestickComponent.prototype.datareader;
    /** @type {?} */
    CandlestickComponent.prototype.data;
    /** @type {?} */
    CandlestickComponent.prototype.level;
    /** @type {?} */
    CandlestickComponent.prototype.target;
    /** @type {?} */
    CandlestickComponent.prototype.drillabledatakey;
    /** @type {?} */
    CandlestickComponent.prototype.labelcolor;
    /** @type {?} */
    CandlestickComponent.prototype.labelflag;
    /** @type {?} */
    CandlestickComponent.prototype.hScale;
    /** @type {?} */
    CandlestickComponent.prototype.chartId;
    /** @type {?} */
    CandlestickComponent.prototype.divid;
    /** @type {?} */
    CandlestickComponent.prototype.onLegendClick;
    /** @type {?} */
    CandlestickComponent.prototype.drillableFlag;
    /** @type {?} */
    CandlestickComponent.prototype.resizeflag;
    /** @type {?} */
    CandlestickComponent.prototype.predefinedColor;
    /** @type {?} */
    CandlestickComponent.prototype.keyArray;
    /** @type {?} */
    CandlestickComponent.prototype.transformeddata;
    /** @type {?} */
    CandlestickComponent.prototype.height;
    /** @type {?} */
    CandlestickComponent.prototype.width;
    /** @type {?} */
    CandlestickComponent.prototype.margin;
    /** @type {?} */
    CandlestickComponent.prototype.x;
    /** @type {?} */
    CandlestickComponent.prototype.y;
    /** @type {?} */
    CandlestickComponent.prototype.svg;
    /** @type {?} */
    CandlestickComponent.prototype.tooltip;
    /** @type {?} */
    CandlestickComponent.prototype.legendArray;
    /** @type {?} */
    CandlestickComponent.prototype.httpresponse;
    /** @type {?} */
    CandlestickComponent.prototype.myservice;
    /** @type {?} */
    CandlestickComponent.prototype.device;
}
