// amexio-chart-d3/esm2015/lib/groupbar/groupbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ElementRef, Output, EventEmitter } from '@angular/core';
import { AmexioD3BaseChartComponent } from "../base/base.component";
import { CommanDataService } from '../services/comman.data.service';
import { DeviceQueryService } from '../services/device.query.service';
import * as d3 from 'd3';
export class GroupbarComponent extends AmexioD3BaseChartComponent {
    /**
     * @param {?} myservice
     * @param {?} device
     */
    constructor(myservice, device) {
        super('multibar');
        this.myservice = myservice;
        this.device = device;
        this.legend = true;
        this.barwidth = 0;
        this.labelcolor = "black";
        this.labelflag = false;
        this.onLegendClick = new EventEmitter();
        this.svgwidth = 300;
        this.svgheight = 300;
        this.level = 0;
        this.drillabledatakey = [];
        this.hScale = true;
        this.drillableFlag = true;
        this.resizeflag = false;
        this.groupbarchartArray = [];
        this.urllegendArray = [];
        this.predefinedcolors = ["#3366cc", "#dc3912", "#ff9900", "#109618", "#990099", "#0099c6", "#dd4477", "#66aa00", "#b82e2e", "#316395", "#994499", "#22aa99", "#aaaa11", "#6633cc", "#e67300", "#8b0707", "#651067", "#329262", "#5574a6", "#3b3eac"];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.level <= 1) {
            /** @type {?} */
            let res;
            if (this.httpmethod && this.httpurl) {
                this.myservice.fetchUrlData(this.httpurl, this.httpmethod).subscribe((response) => {
                    res = response;
                }, (error) => {
                }, () => {
                    setTimeout(() => {
                        this.data = this.getResponseData(res);
                        this.initializeData();
                        this.plotD3Chart();
                    }, 0);
                });
            }
            else if (this.data) {
                setTimeout(() => {
                    this.data = this.getResponseData(this.data);
                    this.initializeData();
                    this.plotD3Chart();
                }, 0);
            }
        }
    }
    /**
     * @param {?} data
     * @return {?}
     */
    fetchData(data) {
        /** @type {?} */
        let requestJson;
        /** @type {?} */
        let key = this.drillabledatakey;
        /** @type {?} */
        let resp;
        if (this.drillabledatakey.length) {
            /** @type {?} */
            let drillabledata = this.getMultipleDrillbleKeyData(data, key);
            requestJson = drillabledata;
        }
        else {
            requestJson = data;
        }
        if (this.httpmethod && this.httpurl) {
            this.myservice.postfetchData(this.httpurl, this.httpmethod, requestJson).subscribe((response) => {
                resp = response;
            }, (error) => {
            }, () => {
                setTimeout(() => {
                    this.data = this.getResponseData(resp);
                    this.drawChart();
                }, 0);
            });
        }
    }
    /**
     * @return {?}
     */
    drawChart() {
        setTimeout(() => {
            this.initializeData();
            this.plotD3Chart();
        }, 0);
    }
    /**
     * @param {?} httpResponse
     * @return {?}
     */
    getResponseData(httpResponse) {
        /** @type {?} */
        let responsedata = httpResponse;
        if (this.datareader != null) {
            /** @type {?} */
            const dr = this.datareader.split('.');
            for (const ir of dr) {
                responsedata = responsedata[ir];
            }
        }
        else {
            responsedata = httpResponse;
        }
        return responsedata;
    }
    /**
     * @return {?}
     */
    plotD3Chart() {
        this.convertToJSON();
        this.plotGroupBarChart();
        this.transformData(this.data);
    }
    /**
     * @return {?}
     */
    plotGroupBarChart() {
        /** @type {?} */
        const tooltip = this.toolTip(d3);
        /** @type {?} */
        let colors = this.predefinedcolors;
        // this.svgwidth = this.chartId.nativeElement.offsetWidth;
        if (this.resizeflag == false) {
            if (this.chartId) {
                this.svgwidth = this.chartId.nativeElement.offsetWidth;
            }
            else {
                this.svgwidth = this.svgwidth;
            }
        }
        /** @type {?} */
        const margin = { top: 20, right: 20, bottom: 50, left: 40 };
        /** @type {?} */
        const width = this.svgwidth - margin.left - margin.right;
        /** @type {?} */
        const height = this.svgheight - margin.top - margin.bottom;
        this.svg = d3.select("#" + this.componentId)
            .attr("width", width + margin.left + margin.right)
            .attr("height", height + margin.top + margin.bottom)
            .append("g")
            .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
        /** @type {?} */
        const x0 = d3.scaleBand()
            .rangeRound([0, width])
            .padding(0.2);
        /** @type {?} */
        const x1 = d3.scaleBand().padding(0.1);
        /** @type {?} */
        const y = d3.scaleLinear()
            .rangeRound([height, 0]);
        //setting x and y domains
        this.years = this.groupbarchartArray.map(function (d) { return d.labels; });
        /** @type {?} */
        let label = this.groupbarchartArray[0].values.map(function (d) { return d.label; });
        x0.domain(this.years);
        x1.domain(label).rangeRound([0, x0.bandwidth()]);
        y.domain([0, d3.max(this.groupbarchartArray, function (labels) { return d3.max(labels.values, function (d) { return d.value; }); })]);
        //dynamic barwidth
        if (this.barwidth > 0) {
            this.barwidth = this.barwidth;
        }
        else {
            this.barwidth = x0.bandwidth;
        }
        // add x axis to svg
        if (this.device.IsDesktop() == true) {
            this.svg.append("g")
                .attr("transform", "translate(0," + height + ")")
                .call(d3.axisBottom(x0));
        }
        else {
            this.svg.append("g")
                .attr("transform", "translate(0," + height + ")")
                .call(d3.axisBottom(x0)).
                selectAll("text")
                .attr("y", 0)
                .attr("x", 9)
                .attr("dy", ".35em")
                .attr("transform", "rotate(60)")
                .style("text-anchor", "start");
        }
        //add y axis to svg
        this.svg.append("g")
            .call(d3.axisLeft(y)
            .ticks(10));
        this.plotLine(this.svg, y, height, width);
        // svg.select('.y').transition().duration(500).delay(1300).style('opacity', '1');
        //adding bars
        /** @type {?} */
        let slice = this.svg.selectAll(".slice")
            .data(this.groupbarchartArray)
            .enter().append("g")
            .attr("class", "g")
            .attr("transform", function (d) { return "translate(" + x0(d.labels) + ",0)"; });
        slice.selectAll("rect")
            .data(function (d) { return d.values; })
            .enter().append("rect")
            .attr("width", x1.bandwidth)
            .attr("x", function (d) {
            return x1(d.label);
        })
            .style("fill", function (d, index) { return colors[index]; })
            .attr("y", function (d) { return y(0); })
            .attr("height", function (d) { return height - y(0); })
            .attr("cursor", "pointer")
            .on("mouseover", (d) => {
            return tooltip.style("visibility", "visible");
        })
            .on("mousemove", (d) => {
            return tooltip.html(this.setKey(d)
            //  this.toolTipContent(d)
            )
                .style("top", (d3.event.pageY - 10) + "px")
                .style("left", (d3.event.pageX + 10) + "px");
        }).on("mouseout", (d) => {
            return tooltip.style("visibility", "hidden");
        })
            .on("click", (d) => {
            this.groupbarClick(d);
            this.fordrillableClick(this, d, event);
            return tooltip.style("visibility", "hidden");
            // this.chartClick(d);
        });
        // -------------------------------------------------------
        if (this.labelflag) {
            slice.selectAll("text")
                .data(function (d) { return d.values; })
                .enter().append("text")
                .attr("width", x1.bandwidth)
                .attr("x", function (d) {
                return x1(d.label) + x1.bandwidth() / 2;
            })
                .attr("y", function (d) { return y(d.value); })
                .attr("height", function (d) { return height - y(0); })
                .style("font-weight", "bold")
                .style("font-size", "1vw")
                .attr("text-anchor", "middle")
                .attr("fill", (d) => {
                if (this.labelcolor.length > 0) {
                    return this.labelcolor;
                }
                else {
                    return "black";
                }
            })
                .text(function (d) {
                return d.value;
            });
        }
        slice.selectAll("rect")
            .attr("y", function (d) {
            return y(d.value);
        })
            .attr("height", function (d) { return height - y(d.value); });
    }
    /**
     * @param {?} d
     * @return {?}
     */
    groupbarClick(d) {
        /** @type {?} */
        let object = {};
        object[d.label] = d.value;
        object[this.xaxisData] = d.xaxis;
        this.chartClick(object);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    resize(event) {
        this.svgwidth = 0;
        this.svg.selectAll("*").remove();
        this.resizeflag = true;
        this.svgwidth = this.divid.nativeElement.offsetWidth;
        this.plotGroupBarChart();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    legendClick(event) {
        // const legendNode = JSON.parse(JSON.stringify(event));
        // delete legendNode.color;
        // const legendNode = JSON.parse(JSON.stringify(event));
        // delete legendNode.color;
        /** @type {?} */
        let obj = {};
        obj["label"] = event.label;
        /** @type {?} */
        let data = [];
        event.data.forEach(element => {
            /** @type {?} */
            let object = {};
            object[element.label] = element.value;
            data.push(object);
        });
        obj["data"] = data;
        this.onLegendClick.emit(obj);
        // this.onLegendClick.emit(legendNode);
    }
    //2d array to json conversion
    /**
     * @return {?}
     */
    convertToJSON() {
        /** @type {?} */
        let groupChartObj = { "labels": "", values: [] };
        /** @type {?} */
        let firstRowOfData = this.data[0];
        this.xaxisData = this.data[0][0];
        for (let i = 1; i < this.data.length; i++) {
            /** @type {?} */
            let multiSeriesArray = [];
            /** @type {?} */
            let valueOfJ;
            for (let j = 1; j < this.data[i].length; j++) {
                valueOfJ = this.data[i][0];
                /** @type {?} */
                let singleBarObj = {};
                singleBarObj["value"] = this.data[i][j];
                singleBarObj["label"] = firstRowOfData[j];
                singleBarObj["xaxis"] = this.data[i][0];
                multiSeriesArray.push(singleBarObj);
            }
            if (multiSeriesArray.length) {
                groupChartObj["values"] = multiSeriesArray;
                /** @type {?} */
                let newLabelsValues = valueOfJ;
                /** @type {?} */
                let newGroupDataObj = Object.assign({}, groupChartObj);
                newGroupDataObj['labels'] = newLabelsValues + '';
                this.groupbarchartArray.push(newGroupDataObj);
            }
        }
    }
    /**
     * @return {?}
     */
    transformto2dArray() {
        /** @type {?} */
        let i = 0;
        /** @type {?} */
        let result = [];
        result.push("years");
        this.data.forEach((element, i) => {
            if (i == 0) {
                element.values.forEach(element2 => {
                    result.push(element2.label);
                });
            }
        });
        this.urllegendArray.push(result);
        /** @type {?} */
        let temparray = [];
        this.data.forEach(element => {
            temparray = [];
            /** @type {?} */
            let values;
            /** @type {?} */
            let year = element.labels;
            temparray.push(year);
            element.values.forEach(element2 => {
                values = element2.value;
                temparray.push(values);
            });
            this.urllegendArray.push(temparray);
        });
    }
    /**
     * @param {?} data
     * @return {?}
     */
    transformData(data) {
        this.keyArray = [];
        this.legendArray = [];
        data.forEach((element, i) => {
            if (i == 0) {
                element.forEach((innerelement, index) => {
                    if (index > 0) {
                        this.legendArray[innerelement] = { 'data': [] };
                        this.keyArray.push(innerelement);
                    }
                });
            }
        });
        data.forEach((element, index) => {
            if (index > 0) {
                /** @type {?} */
                let obj = {};
                element.forEach((innerelement, innerindex) => {
                    if (innerindex >= 0) {
                        /** @type {?} */
                        const key = this.keyArray[innerindex - 1];
                        obj[key] = element[innerindex];
                        /** @type {?} */
                        const legenddata = this.legendArray[key];
                        if (legenddata) {
                            legenddata.data.push({ 'value': element[innerindex], 'label': element[0] });
                        }
                    }
                });
            }
        });
        this.legends = [];
        this.keyArray.forEach((element, index) => {
            /** @type {?} */
            const legenddata = this.legendArray[element];
            /** @type {?} */
            let object = { 'label': element, 'color': this.predefinedcolors[index], 'data': legenddata.data };
            this.legends.push(object);
        });
    }
    /**
     * @param {?} d
     * @return {?}
     */
    setKey(d) {
        /** @type {?} */
        let object = {};
        object[d.label] = d.value;
        object[this.xaxisData] = d.xaxis;
        return (this.toolTipForBar(object));
    }
    /**
     * @param {?} g
     * @param {?} y
     * @param {?} height
     * @param {?} width
     * @return {?}
     */
    plotLine(g, y, height, width) {
        if (this.hScale) {
            g.append('g')
                .attr("color", "lightgrey")
                .call(d3.axisLeft(y)
                .tickSize(-width).tickFormat(''));
        }
    }
}
GroupbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'amexio-d3-chart-multiseries',
                template: `<div   *ngIf="drillableFlag" #divid (window:resize)="resize($event)"> 
<table style="width: 100%;" #chartId>
    <tr>
        <td colspan="2" style="font-size: 18px; font-weight: bold; "> {{title}} </td>
    </tr>
    <tr>
        <td>
            <svg [attr.height]="height" [attr.width]="width" [attr.id]="componentId">
            </svg>
        </td>
    </tr>
    <tr>
        <td style="display: block; padding: 5px;">
            <amexio-d3-legend *ngIf="legend" [horizontal]="true" (onClick)="legendClick($event)" [data]="legends" style="height:100%;"></amexio-d3-legend>
        </td>
    </tr>
</table>
</div>


`,
                styles: [``]
            },] },
];
GroupbarComponent.ctorParameters = () => [
    { type: CommanDataService },
    { type: DeviceQueryService }
];
GroupbarComponent.propDecorators = {
    chartId: [{ type: ViewChild, args: ['chartId',] }],
    divid: [{ type: ViewChild, args: ['divid',] }],
    drillid: [{ type: ViewChild, args: ['drillid',] }],
    data: [{ type: Input, args: ['data',] }],
    legend: [{ type: Input }],
    barwidth: [{ type: Input }],
    labelcolor: [{ type: Input, args: ['label-color',] }],
    labelflag: [{ type: Input, args: ['label',] }],
    onLegendClick: [{ type: Output }],
    svgwidth: [{ type: Input, args: ['width',] }],
    svgheight: [{ type: Input, args: ['height',] }],
    datareader: [{ type: Input, args: ['data-reader',] }],
    level: [{ type: Input, args: ['level',] }],
    target: [{ type: Input, args: ['target',] }],
    drillabledatakey: [{ type: Input, args: ['drillable-data',] }],
    hScale: [{ type: Input, args: ['horizontal-scale',] }]
};
if (false) {
    /** @type {?} */
    GroupbarComponent.prototype.chartId;
    /** @type {?} */
    GroupbarComponent.prototype.divid;
    /** @type {?} */
    GroupbarComponent.prototype.drillid;
    /** @type {?} */
    GroupbarComponent.prototype.data;
    /** @type {?} */
    GroupbarComponent.prototype.legend;
    /** @type {?} */
    GroupbarComponent.prototype.barwidth;
    /** @type {?} */
    GroupbarComponent.prototype.labelcolor;
    /** @type {?} */
    GroupbarComponent.prototype.labelflag;
    /** @type {?} */
    GroupbarComponent.prototype.onLegendClick;
    /** @type {?} */
    GroupbarComponent.prototype.svgwidth;
    /** @type {?} */
    GroupbarComponent.prototype.svgheight;
    /** @type {?} */
    GroupbarComponent.prototype.datareader;
    /** @type {?} */
    GroupbarComponent.prototype.level;
    /** @type {?} */
    GroupbarComponent.prototype.target;
    /** @type {?} */
    GroupbarComponent.prototype.drillabledatakey;
    /** @type {?} */
    GroupbarComponent.prototype.hScale;
    /** @type {?} */
    GroupbarComponent.prototype.drillableFlag;
    /** @type {?} */
    GroupbarComponent.prototype.resizeflag;
    /** @type {?} */
    GroupbarComponent.prototype.groupbarchartArray;
    /** @type {?} */
    GroupbarComponent.prototype.legendArray;
    /** @type {?} */
    GroupbarComponent.prototype.xaxisData;
    /** @type {?} */
    GroupbarComponent.prototype.keyArray;
    /** @type {?} */
    GroupbarComponent.prototype.legends;
    /** @type {?} */
    GroupbarComponent.prototype.years;
    /** @type {?} */
    GroupbarComponent.prototype.urllegendArray;
    /** @type {?} */
    GroupbarComponent.prototype.svg;
    /** @type {?} */
    GroupbarComponent.prototype.myservice;
    /** @type {?} */
    GroupbarComponent.prototype.device;
}
